### PR TITLE
[lldb] Add early CMake check for 'make' tool

### PR DESCRIPTION
**Pull request by weliveindetail**
Around 400 of LLDB's dotest.py based tests require the `make` tool to be found in Path. If it's not found, they fail with an obscure error and show up as `UNRESOLVED`. llvm-lit takes care of MSYS based testing tools like cat, printf, etc., but make is not part of that. Let's catch the situation early and raise an error if LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS was enabled. This error is not fatal: It should fail the build, but not immediately stop the configuration process. There might be other issues further down the line that can be caught in the same buildbot run.

### CODE REVIEW

**Review by JDevlieghere — COMMENTED** [review]
I like the idea of looking for Make at configuration time. However, I think if we go that route, we should pass make as an explicit argument to `dotest.py`:

 - We do for other tools that we find at configuration time (such as `dsymutil`).
 - It eliminates the possibility of `dotest.py` finding a different `make` than CMake. 

Looking at `dotest.py`, it already supports `--make` so hopefully this is trivial to hook up.

**Comment by bulbazord** [comment]
+1 to what Jonas said. Making `dotest.py` more explicit makes reproducing test behavior more reliable. :)

**Comment by weliveindetail** [comment]
Thanks for the quick feedback. Yes, makes sense. Here we go.

File: lldb/test/API/lit.cfg.py
Comment (weliveindetail) [inline lldb/test/API/lit.cfg.py:255]: `--make` is added if `LLDB_TEST_MAKE` is defined explicitly or `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS=ON`

File: lldb/packages/Python/lldbsuite/test/dotest.py
Comment (weliveindetail) [inline lldb/packages/Python/lldbsuite/test/dotest.py:276]: The default install path on Windows is `C:\Program Files (x86)\GnuWin32\bin\make.exe` unfortunately.

**Review by weliveindetail — COMMENTED** [review]
Two notes inline

File: lldb/test/CMakeLists.txt
Comment (JDevlieghere) [inline lldb/test/CMakeLists.txt:39]: Is there a reason to conditionalize this on `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS`? I would hoist this up so that we always have CMake configure `make` and only conditionalize the warning.

File: lldb/test/CMakeLists.txt
Comment (JDevlieghere) [inline lldb/test/CMakeLists.txt:42]: TIL about `SEND_ERROR`. I think it would be nice to use that for the rest of the strict requirements, but obviously that's outside the scope of this PR.

**Review by JDevlieghere — APPROVED** [review]
LGTM with detection hoisted.

File: lldb/test/CMakeLists.txt
Comment (JDevlieghere) [reply lldb/test/CMakeLists.txt:42]: https://github.com/llvm/llvm-project/pull/111531

Comment (JDevlieghere) [reply lldb/test/CMakeLists.txt:42]: https://github.com/llvm/llvm-project/pull/111729

File: lldb/test/CMakeLists.txt
Comment (weliveindetail) [reply lldb/test/CMakeLists.txt:39]: Well, no there is no GOOD reason. For the moment it was easy to keep it guarded by `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS` since it is `OFF` by default and won't fall on people's feet immediately. However, I am afraid we should special-case NetBSD and FreeBSD here to look for `gmake` instead of `make`.

Let me fix this in a follow-up PR that includes the author of https://github.com/llvm/llvm-project/pull/93883

File: lldb/packages/Python/lldbsuite/test/dotest.py
Comment (labath) [reply lldb/packages/Python/lldbsuite/test/dotest.py:276]: That may be fine, but I think this is the wrong place to quote it. I think it should happen somewhere around the place where this variable is concatenated into a string (or passed to whatever it is that requires it to be quoted).

That said, after tracing this variable, I'm a little unsure as to why is this needed. IIUC, this eventually makes its way to the `subprocess.check_output` call [here](https://github.com/llvm/llvm-project/blob/f1eac7761704fa4e16c2619a84890baad380fdce/lldb/packages/Python/lldbsuite/test/lldbtest.py#L1499), which should be able to handle quoting on its own. 

Can you explain what the problem was without this?

File: lldb/test/CMakeLists.txt
Comment (labath) [reply lldb/test/CMakeLists.txt:39]: How about doing it in the other order? I think it would be somewhat strange if `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS` actually changed what is being used (instead of just erroring out when detecting errors).

That said, I think that:
- this PR is small enough to do both of those things together
- all that's needed is to do something like `find_program(LLDB_DEFAULT_TEST_MAKE NAMES gmake make)`. `gmake` can exist on non-netbsd systems as well, and it should not be an error to use that (in fact, `make` is a symlink to `gmake` on my system)

Comment (weliveindetail) [reply lldb/test/CMakeLists.txt:39]: @labath Thanks for your feedback! I already set up the follow-up. I will take your proposal into account. Hope it's ok to do it there: https://github.com/llvm/llvm-project/pull/111744

Comment (weliveindetail) [reply lldb/test/CMakeLists.txt:39]: Yes, in the end it's small enough indeed. Let me abandon https://github.com/llvm/llvm-project/pull/111744 and add these changes here.

File: lldb/packages/Python/lldbsuite/test/dotest.py
Comment (weliveindetail) [reply lldb/packages/Python/lldbsuite/test/dotest.py:276]: I added this because yesterday my tests failed with:
```
'C:/Program' is not recognized as an internal or external command,
operable program or batch file.
```

I double-checked again today and this doesn't happen anymore. I am not sure why, maybe I had an inconsistent configuration state. The parameter for dotest.py is `--make C:/Program Files (x86)/GnuWin32/bin/make.exe` and the command that lldbtest.py runs is starting with `C:/Program Files (x86)/GnuWin32/bin/make.exe VPATH=...`. I cleared the cached in `lldb-test-build.noindex` and checked sure the binaries are recreated successfully by the test suite. It works without the quotes.

I am attaching a log file [TestBreakpointByLineAndColumn.log](https://github.com/user-attachments/files/17326545/TestBreakpointByLineAndColumn.log) with the commands dumped for future reference. I will the quoting and assume it just keeps working.

File: lldb/test/API/lit.cfg.py
Comment (weliveindetail) [reply lldb/test/API/lit.cfg.py:255]: Changed in follow-up patch https://github.com/llvm/llvm-project/pull/111531/commits/3f6b93d932d9755c668d31b9c75c152bb3599bbe We now always detect `make`.

**Comment by zeroomega** [comment]
Hi,

I noticed that our LLDB Windows builders all failed after this change is landed:

```
CMake Error at C:/b/s/w/ir/x/w/llvm-llvm-project/lldb/test/API/CMakeLists.txt:60 (message):
   LLDB tests require 'make' tool.  Please pass via `LLDB_TEST_MAKE` (or
   otherwise disable tests with `LLDB_INCLUDE_TESTS=OFF`)
```

We are aware our windows bots don't have make tool and we are not planning to add it anyway. Not all lldb tests require `make` to run and it feels wrong to me to mandate `make` tool for a few tests that requires it. Could it be changed to disable such tests when make is not present instead of throw an error to prevent the lldb build?

**Comment by JDevlieghere** [comment]
> Not all lldb tests require `make` to run and it feels wrong to me to mandate `make` tool for a few tests that requires it. 

All the API tests build their inferior with Make and looking at the last successful run on Windows, we have 1206 API tests, 561 shell tests and 261 unit tests. Even if we only consider the ones that are passing, the API tests are still the vast majority. I'm not sure how you come to the conclusion that only "a few tests" require it?

**Comment by weliveindetail** [comment]
In fact, I ran into that on Friday myself. I checked `LLDB_INCLUDE_TESTS=Off` in my setup and it appeared to be incompatible with `LLDB_ENABLE_PYTHON=On` if `lldb-python` is part of `LLVM_DISTRIBUTION_COMPONENTS`. The error said something like `lldb-python-scripts` is lacking an `install-stripped` target. I couldn't fix that quickly and just wired up the `LLDB_TEST_MAKE` config.

@zeroomega Any similar experiences with `LLDB_INCLUDE_TESTS=Off`?

**Comment by zeroomega** [comment]
> In fact, I ran into that on Friday myself. I checked `LLDB_INCLUDE_TESTS=Off` in my setup and it appeared to be incompatible with `LLDB_ENABLE_PYTHON=On` if `lldb-python` is part of `LLVM_DISTRIBUTION_COMPONENTS`. The error said something like `lldb-python-scripts` is lacking an `install-stripped` target. I couldn't fix that quickly and just wired up the `LLDB_TEST_MAKE` config.
> 
> @zeroomega Any similar experiences with `LLDB_INCLUDE_TESTS=Off`?

No, I didn't see this issue. Our build passed after I set `LLDB_INCLUDE_TESTS=Off` on our Windows bots.  But we also didn't set `lldb-python` in `LLVM_DISTRIBUTION_COMPONENTS`. We do have `lldb-python-scripts`. You can see our cmake invocation here: 
https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8734080863088320417/+/u/build_and_install_lldb/configure/l_execution_details

**Comment by GkvJwa** [comment]
In fact, for many developers on Windows, `make.exe` will not be in the environment PATH, which will cause cmake to fail to find and ultimately delay the build of the project error.
I personally think it is better to use `FATAL_ERROR` instead of `SEND_ERROR`. After all, it will cause cmake to continue executing for a short period of time and finally display

```
CMake Error at C:/llvm-project/lldb/test/API/CMakeLists.txt:60 (message):
  LLDB tests require 'make' tool.  Please pass via `LLDB_TEST_MAKE` (or
  otherwise disable tests with `LLDB_INCLUDE_TESTS=OFF`)


-- BugpointPasses ignored -- Loadable modules not supported on this platform.
-- Google Benchmark version: v0.0.0, normalized to 0.0.0
-- Looking for shm_open in rt
-- Looking for shm_open in rt - not found
-- Performing Test HAVE_CXX_FLAG_EHS_
-- Performing Test HAVE_CXX_FLAG_EHS_ - Success
-- Performing Test HAVE_CXX_FLAG_EHA_
-- Performing Test HAVE_CXX_FLAG_EHA_ - Success
-- Compiling and running to test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Compiling and running to test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- failed to compile
CMake Warning at C:/llvm-project/third-party/benchmark/CMakeLists.txt:319 (message):
  Using std::regex with exceptions disabled is not fully supported


-- Compiling and running to test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK -- success
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
-- Found Threads: TRUE
-- Compiling and running to test HAVE_PTHREAD_AFFINITY
-- Performing Test HAVE_PTHREAD_AFFINITY -- failed to compile
-- Configuring incomplete, errors occurred!
```

**Comment by weliveindetail** [comment]
I think it's valuable to get the full config log on buildbots. If this check becomes too much of a burden, I'd rather propose to either turn the error into a warning or revisit the option to guard it by `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS` (see my original patch for details). What about that? @GkvJwa

**Comment by GkvJwa** [comment]
> I think it's valuable to get the full config log on buildbots. If this check becomes too much of a burden, I'd rather propose to either turn the error into a warning or revisit the option to guard it by `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS` (see my original patch for details). What about that? @GkvJwa

Personally, warnings are better than errors for developers

**Comment by weliveindetail** [comment]
Yes, for devs it might be too much of a burden and eventually bot maintainers should be aware. If we want the error behavior back, we could make it conditional on `LLDB_ENFORCE_STRICT_TEST_REQUIREMENTS`. For now I didn't want to make it more complicated than necessary.

**Pull request merged**
